### PR TITLE
Added inline_assembly Task (Run .NET assemblies in disposable AppDomains)

### PR DESCRIPTION
**Pull request by thiagomayllart**
Hi! I've added a new file DarkMelkor.cs in the Utils folder, while also modifying the AssemblyManager.cs file to add the inline_assembly function.

Some months ago, @b33f (FuzzySecurity) developed Melkor: https://github.com/FuzzySecurity/Sharp-Suite/. This tool uses CryptProtectData and CryptUnprotectData functions to encrypt and decrypt .NET assemblies in memory, which is a good way to avoid memory scans flagging registered assemblies in the Apollo agent. Melkor also leverages the capability to run these loaded assemblies in disposable AppDomains, which is a good alternative to fork&run tasks in case process injection (at least through some techniques) is not an option.

The only issue with Melkor is the fact that it is not capable of running these Assemblies in case you are doing this in an injected process. Because of that, some days ago I've modified Melkor into this version: https://github.com/thiagomayllart/DarkMelkor.

This version still encrypts and decrypts the .NET assemblies through the crypt32 functions. However, the way it runs is based in this article: https://www.accenture.com/us-en/blogs/cyber-defense/clrvoyance-loading-managed-code-into-unmanaged-processes.

Bryan Alexander and Josh Stone found a way to "bypass" the way the AppDomains segregate the .NET assemblies. They found that it is possible to create two Cross App Domain Delegates: one of them pointing to a function that can be resolved by our disposable AppDomain (generally a function in the mscorlib) and another one pointing to the malicious function (which, in the context of the Apollo agent, will be the one actually invoking the loaded bytes of our registered assembly). 

By patching the initial bytes of the function (the one that can be resolved) with the jmp instructions (mov rax, &delegate; jmp rax) to the malicious delegate, it is possible to callback this function in a way that, instead of running the non-malicious one, it will end up jumping to the address of the other one, thus, loading the malicious code.

I've also modified the other functions that handle .NET assemblies in AssemblyManager.cs in order to decrypt the registered assemblies. This might help to avoid some memory scans even when using the fork&run tasks.

Some things to mention: 

Since Apollo is compiled for .NET 4.0, I have not added the correct position to patch in case of .NET 4.5 and above. In case that's necessary, i've added these in here: https://github.com/thiagomayllart/DarkMelkor/blob/main/DarkMelkor/DarkMelkor/DarkMelkor/hDarkMelkor.cs (line 126). It seems that in .NET 4.5, the patch position when you are doing this in an injected process is different from when you are doing it by running the compiled Apollo.exe directly. I have not verified these positions in >.NET 4.5.

Currently, the way DarkMelkor modifies the memory regions to RWX is by using the VirtualProtect function. It also leverages RtlZeroMemory by declaring an extern. These problems might be solved by adding ntprotectvirtualmemory and the capability to resolve syscalls dynamically (without dinvoke/without reading ntdll from disk), which solves the problem of having an EDR hooking these functions. I intend to add these in a further request.

Credit goes specially to @b33f (Fuzzy Security), Bryan Alexander and Josh Stone, who actually researched the mentioned techniques, I've just assembled them.

### CODE REVIEW

File: Payload_Type/apollo/agent_code/Apollo/Utils/DarkMelkor.cs [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+        }
+
+        [Flags]
+        public enum AllocationType : uint
+        {
+            Commit = 0x1000,
+            Reserve = 0x2000,
+            Decommit = 0x4000,
+            Release = 0x8000,
+            Reset = 0x80000,
```
Comment (djhohnstein) [inline Payload_Type/apollo/agent_code/Apollo/Utils/DarkMelkor.cs:85]: All these definitions should exist in the correct file under the Native folde/namespace. There should be things for methods, enums, constants, etc.

File: Payload_Type/apollo/mythic/agent_functions/inline_assembly.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    is_file_browse = False
+    is_process_list = False
+    is_download_file = False
+    is_upload_file = False
+    is_remove_file = False
+    author = "@thiagomayllart"
+    argument_class = InlineAssemblyArguments
+    attackmapping = ["T1547"]
+
+    async def create_tasking(self, task: MythicTask) -> MythicTask:
```
Comment (djhohnstein) [inline Payload_Type/apollo/mythic/agent_functions/inline_assembly.py:41]: Set the display params to something nice

File: Payload_Type/apollo/agent_code/Apollo/CommandModules/AssemblyManager.cs [hunk 3]
```diff
@@ -126,27 +134,73 @@ private static void DeleteAssembly(string key)
...
             {
                 asm = loadedAssemblies[key];
             }
             return asm;
         }
 
+#if INLINE_ASSEMBLY
+        static void InlineAssembly(Job job, Agent implant)
+        {
+            Task task = job.Task;
```
Comment (djhohnstein) [inline Payload_Type/apollo/agent_code/Apollo/CommandModules/AssemblyManager.cs:161]: I'm reasonably certain you need to add a locking mechanism for this task. From the DarkMelkor code, you're modifying the std handles, and if two inline tasks execute concurrently, we'll lose the original handles to the process (not good). By implementing a lock on a static object we ensure task concurrency doesn't stomp on each other, and we simply accept this is a limitation of this function.

**Review by djhohnstein — CHANGES_REQUESTED** [review]
tldr;

1. Change task.display_params to something nice
2. Change the inline_assembly command in the agent to implement a locking mechanism so concurrent executions don't lose the stdout handle forever.
3. Move native api calls into the requisite file under the Native folder and use that namespace instead.

**Comment by djhohnstein** [comment]
** Also please target the dev branch for your merge.

File: Payload_Type/apollo/agent_code/Apollo/CommandModules/AssemblyManager.cs (hunk 3 above)
Comment (thiagomayllart) [reply Payload_Type/apollo/agent_code/Apollo/CommandModules/AssemblyManager.cs]: Added locking mechanism through the variable "inline_assembly_Locker".

File: Payload_Type/apollo/agent_code/Apollo/Utils/DarkMelkor.cs (hunk 1 above)
Comment (thiagomayllart) [reply Payload_Type/apollo/agent_code/Apollo/Utils/DarkMelkor.cs]: Moved the correspondent P/Invoke and Enums to the correct files.

**Comment by thiagomayllart** [comment]
Something necessary to mention is another limitation of this functionality: if the .NET assembly tasked to execute contains an Environment.Exit call, it will likely terminate the agent. It is necessary to remove the call from the code to prevent this behavior.

**Comment by thiagomayllart** [comment]
Updated to this pull request: https://github.com/MythicAgents/Apollo/pull/60
